### ISSUE TITLE
Getting the docker images smaller

### ISSUE THREAD

**Issue by benizar** [siose-innova/siose-docker-library#1]
All these images are pretty functional, however they weight almost 1 Gb after they're built. Then, we still load some SIOSE data (from 100 Mb to >10 Gb). I propose two improvements that could solve this:

1. The [GDAL Dockerfile](https://github.com/siose-innova/siose-docker-library/blob/master/lib/gdal/Dockerfile) is a good example of what we could do if we base our images in Alpine.
2. We also could create **test and production** images. Production images could have several libraries removed after installing extensions.

I don't think this is urgent right now, but it will when other users start using these images.

Any other ideas on reducing the image sizes? Maybe creating other intermediate layers?

**Comment by quommit** [comment]
I'm planning two pull requests for both, a PostgreSQL and a PostGIS image based on Alpine Linux. Those may serve as intermediate layers, as in the following tree:  
  
```
siose-innova/postgres
 |_ siose-innova/postgis
      |_ siose-innova/postgis-ext
```  
  
This way, dependencies for building extensions get integrated in image `siose-innova/postgis-ext`. The problem is that those dependencies must be removed from final extension images during compilation. Alpine's package manager (APK) provides a clean and safe way to do it so that you end up with a lighter binary (no gcc, make, git, certificates or whatever). The point is, I've never removed build dependencies coming from a base image, that is, my build dependencies are defined in the current Dockerfile and removed in the current Dockerfile.  
For instance, you build `siose-innova/pg-landmetrics` (*A*) using `FROM siose-innova/postgis-ext` (*B*). Your dependencies for building extensions are declared in Dockerfile *B* with `apk add --no-cache --virtual .build-deps [here_goes_the_list_of_packages]`.  Then, in Dockerfile *A* you remove them with `apk del .build-deps`. I don't know if the *virtual* package `.build-deps` is reachable after image compilation. In case it were, you are referring to a *virtual* package which was defined in another Dockerfile which, in turn, may have been modified.  
To summarize, the way to go (as far as I'm aware of) is whenever you need to build an image for your PostgreSQL extension, you install/remove your build dependencies (curl, git, gcc, make, etc.) within the scope of that particular image's Dockerfile. So, instead of using this dependency chain:  
  
```
siose-innova/postgres
 |_ siose-innova/postgis
      |_ siose-innova/postgis-ext
           |_ siose-innova/pg-landmetrics
```  
  
you just go for:  
  
```
siose-innova/postgres
 |_ siose-innova/postgis
      |_ siose-innova/pg-landmetrics
```

**Comment by benizar** [comment]
Ok, it seems pretty simple now :)

I will wait till you propose alpine versions for these images.

**Comment by benizar** [comment]
@quommit
Alpine images change very fast... For example, one week later the GDAL image can't be built. Any way to manage this by modifying the dockerfile or we shouldn't build/push alpine images so frecuently?

```
WARNING: This apk-tools is OLD! Some packages might not function properly.
ERROR: unsatisfiable constraints:
  so:libproj.so.12 (missing):
    required by:
                 libspatialite-4.3.0a-r0[so:libproj.so.12]
                 libspatialite-4.3.0a-r0[so:libproj.so.12]
                 libspatialite-4.3.0a-r0[so:libproj.so.12]
```

**Comment by benizar** [comment]
In [this post](https://tuhrig.de/flatten-a-docker-container-or-image/) Thomas Uhrig explains how a docker image of 7GB becomes an image of 3GB. He uses 

```
docker export <CONTAINER ID> | docker import - some-image-name:latest
```

This removes the docker history, but it can be an option for production images.

**Comment by benizar** [comment]
It doesn't seems related to the alpine distribution (3.7 or edge)... Maybe it is needed to do `apk update`?
